### PR TITLE
Fix/543 default branch name rebase on develop

### PR DESCRIPTION
**Pull request by progfolio**
@ahyatt: I've rebased your changes for #543 on top of the current develop branch and squashed them. There were merge conflicts due to work that had been completed on develop while you were working on this which I resolved manually. Please take a look at the diff and make sure I haven't broken/missed anything.

### CODE REVIEW

**Comment by progfolio** [comment]
CI failing with:

```
Wrong type argument: arrayp, nil
```
 during the smoke test.

Will take another look at this later tonight.

File: straight.el [hunk 1]
```diff
@@ -2202,13 +2212,15 @@ Return non-nil. If no local repository, do nothing and return non-nil."
...
+            (branch
+             (or branch
+                 (straight-vc-git--default-remote-branch remote local-repo))))
         (while t
           (while (not (straight-vc-git--ensure-local recipe)))
           (let ((ref (format "%s/%s" remote branch)))
             (when (straight--check-call
-                   "git" "merge-base" "--is-ancestor"
-                   branch ref)
+                   "git" "merge-base" "--is-ancestor" branch ref)
```
Comment (ahyatt) [inline straight.el:2223]: You probably should revert this meaningless whitespace change.

File: straight.el [hunk 2]
```diff
@@ -2235,7 +2247,8 @@ Return non-nil. If no local repository, do nothing and return non-nil."
                             (straight--call
                              "git" "push" remote
                              (format
-                              "refs/heads/%s:refs/heads/%s" branch branch))))
+                              "refs/heads/%s:refs/heads/%s" branch
+                              branch))))
```
Comment (ahyatt) [inline straight.el:2251]: Another whitespace-only change that should be reverted.

**Comment by ahyatt** [comment]
The change from the rebase look fine to me, thanks for doing this.  I made a few minor comments.

**Comment by progfolio** [comment]
Thank you for taking a look at this. I've addressed the whitespace changes.
Unfortunately it looks like we're failing to bootstrap due to a `:branch` value of `nil` for the melpa repo:

<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :preserve t
  :user-dir "543-default-branch-name.straight"
  :pre-bootstrap 
  (setq straight-repository-user "progfolio")
  (setq straight-repository-branch "fix/543-default-branch-name-rebase-on-develop")
  :post-bootstrap 
  (straight-use-package 'wanderlust))
```
</details>

- Test run at: `2020-11-14 00:15:17`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> develop, origin/develop) 7b5a930 2020-11-13`
- emacs-version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, X toolkit, cairo version 1.17.3, Xaw3d scroll bars)
 of 2020-11-13`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe → Cloning melpa...

Debugger entered--Lisp error: (wrong-type-argument arrayp nil)
  replace-regexp-in-string("[^-0-9a-zA-Z_./\n]" "\\\\\\&" nil)
  shell-quote-argument(nil)
  mapconcat(shell-quote-argument ("git" "clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep...") " ")
  straight--process-run("git" "clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep...")
  apply(straight--process-run "git" ("clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep..."))
  straight--call("git" "clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep...")
  apply(straight--call "git" ("clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep..."))
  straight--get-call-raw("git" "clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep...")
  apply(straight--get-call-raw "git" ("clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep..."))
  straight--get-call("git" "clone" "--origin" "origin" "--branch" nil "--no-checkout" "https://github.com/melpa/melpa.git" "/tmp/543-default-branch-name.straight/straight/rep...")
  straight-vc-git--clone-internal(:depth full :remote "origin" :url "https://github.com/melpa/melpa.git" :repo-dir "/tmp/543-default-branch-name.straight/straight/rep..." :branch nil :commit nil)
  straight-vc-git-clone((:type git :host github :repo "melpa/melpa" :no-build t :package "melpa" :local-repo "melpa") nil)
  apply(straight-vc-git-clone ((:type git :host github :repo "melpa/melpa" :no-build t :package "melpa" :local-repo "melpa") nil))
  straight-vc(clone git (:type git :host github :repo "melpa/melpa" :no-build t :package "melpa" :local-repo "melpa") nil)
  straight-vc-clone((:type git :host github :repo "melpa/melpa" :no-build t :package "melpa" :local-repo "melpa"))
  straight--clone-repository((:type git :host github :repo "melpa/melpa" :no-build t :package "melpa" :local-repo "melpa") "Looking for gnu-elpa-mirror recipe")
  #f(compiled-function () #<bytecode -0x949547444f14c1d>)()
  straight--transaction-exec(use-package-\(:type\ git\ :host\ github\ :repo\ \"melpa/melpa\"\ :no-build\ t\ :package\ \"melpa\"\ :local-repo\ \"melpa\"\)-nil-nil :now #f(compiled-function () #<bytecode -0x949547444f14c1d>))
  straight-use-package(melpa nil nil "Looking for gnu-elpa-mirror recipe")
  straight-recipes(retrieve melpa "Looking for gnu-elpa-mirror recipe" gnu-elpa-mirror)
  straight-recipes-retrieve(gnu-elpa-mirror nil)
  straight--convert-recipe((gnu-elpa-mirror :type git :host github :repo "emacs-straight/gnu-elpa-mirror" :no-build t) nil)
  straight-use-package((gnu-elpa-mirror :type git :host github :repo "emacs-straight/gnu-elpa-mirror" :no-build t) #f(compiled-function (package available) #<bytecode -0x17531cd25f1994bf>))
  straight-use-package-lazy((gnu-elpa-mirror :type git :host github :repo "emacs-straight/gnu-elpa-mirror" :no-build t))
  straight-use-recipes((gnu-elpa-mirror :type git :host github :repo "emacs-straight/gnu-elpa-mirror" :no-build t))
  (if straight-recipes-gnu-elpa-use-mirror (straight-use-recipes '(gnu-elpa-mirror :type git :host github :repo "emacs-straight/gnu-elpa-mirror" :no-build t)) (straight-use-recipes (cons 'gnu-elpa (cons ':type (cons 'git (cons ':repo (cons straight-recipes-gnu-elpa-url '...)))))))
  eval-buffer(#<buffer  *load*> nil "/tmp/543-default-branch-name.straight/straight/rep..." nil t)  ; Reading at buffer position 4534
  load-with-code-conversion("/tmp/543-default-branch-name.straight/straight/rep..." "/tmp/543-default-branch-name.straight/straight/rep..." nil t)
  load("/tmp/543-default-branch-name.straight/straight/rep..." nil nomessage)
  (let ((bootstrap-file (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory)) (bootstrap-version 5)) (unless (file-exists-p bootstrap-file) (with-current-buffer (url-retrieve-synchronously (concat "https://raw.githubusercontent.com/" "raxod502/straight.el/develop/install.el") 'silent 'inhibit-cookies) (goto-char (point-max)) (eval-print-last-sexp))) (load bootstrap-file nil 'nomessage))
  (let ((straight-bug-report-args '(:post-bootstrap (progn (straight-use-package 'wanderlust)) :pre-bootstrap (progn (setq straight-repository-user "progfolio") (setq straight-repository-branch "fix/543-default-branch-name-rebase-on-develop")) :user-dir "/tmp/543-default-branch-name.straight" :preserve t :executable "/usr/bin/emacs"))) (setq user-emacs-directory "/tmp/543-default-branch-name.straight") (setq straight-repository-branch "develop") (setq debug-on-error t) (setq straight-repository-user "progfolio") (setq straight-repository-branch "fix/543-default-branch-name-rebase-on-develop") (defvar bootstrap-version) (let ((bootstrap-file (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory)) (bootstrap-version 5)) (unless (file-exists-p bootstrap-file) (with-current-buffer (url-retrieve-synchronously (concat "https://raw.githubusercontent.com/" "raxod502/straight.el/develop/install.el") 'silent 'inhibit-cookies) (goto-char (point-max)) (eval-print-last-sexp))) (load bootstrap-file nil 'nomessage)) (condition-case nil (message "Test run with version: %s" (straight-version)) (error nil)) (straight-use-package 'wanderlust))
  eval((let ((straight-bug-report-args '(:post-bootstrap (progn (straight-use-package ...)) :pre-bootstrap (progn (setq straight-repository-user "progfolio") (setq straight-repository-branch "fix/543-default-branch-name-rebase-on-develop")) :user-dir "/tmp/543-default-branch-name.straight" :preserve t :executable "/usr/bin/emacs"))) (setq user-emacs-directory "/tmp/543-default-branch-name.straight") (setq straight-repository-branch "develop") (setq debug-on-error t) (setq straight-repository-user "progfolio") (setq straight-repository-branch "fix/543-default-branch-name-rebase-on-develop") (defvar bootstrap-version) (let ((bootstrap-file (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory)) (bootstrap-version 5)) (unless (file-exists-p bootstrap-file) (with-current-buffer (url-retrieve-synchronously (concat "https://raw.githubusercontent.com/" "raxod502/straight.el/develop/install.el") 'silent 'inhibit-cookies) (goto-char (point-max)) (eval-print-last-sexp))) (load bootstrap-file nil 'nomessage)) (condition-case nil (message "Test run with version: %s" (straight-version)) (error nil)) (straight-use-package 'wanderlust)) t)
  command-line-1(("--eval" "(let\n    ((straight-bug-report-args\n      '(:post-..."))
  command-line()
  normal-top-level()


```
</details>

**Comment by progfolio** [comment]
@ahyatt: OK, I think we're good now. The nil branch was a result of a change made in #608 that hadn't been accounted for.
@raxod502: Just want to make sure this won't cause a regression as far as the behavior of #607 is concerned.

https://github.com/raxod502/straight.el/pull/630/commits/e877d793e51ef65eaec3bc87dbd1c7ee3da8f8ea

look good to you?


So long as the above is good, it looks like it's working well.
To test I created a branch with #549 rebased on top of these changes and tried building one of the el-get recipes which was failing due to having a default branch of "trunk":

<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :user-dir "543-el-get-test.straight"
  :pre-bootstrap 
  (setq straight-repository-user "progfolio")
  (setq straight-repository-branch "test/543-with-el-get-support")
  :post-bootstrap 
  (setq straight-recipe-repositories
        '(el-get melpa))
  (straight-use-package
   '(company-emoji :source el-get :post-build
                   (message "%S"
                            (shell-command-to-string "git branch")))))
```
</details>

- Test run at: `2020-11-14 00:57:32`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> test/543-with-el-get-support, fork/test/543-with-el-get-support) 5a2e6a4 2020-11-14`
- emacs-version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, X toolkit, cairo version 1.17.3, Xaw3d scroll bars)
 of 2020-11-13`

<details><summary>Output</summary>

```emacs-lisp
Test run with version: prerelease (HEAD -> test/543-with-el-get-support, origin/test/543-with-el-get-support) 5a2e6a4 2020-11-14
Building company-emoji...

"* trunk
"
Building company-emoji...done

```
</details>

**Comment by raxod502** [comment]
> Just want to make sure this won't cause a regression as far as the behavior of #607 is concerned

Yeah, I think this is alright. If the branch is nil, then we want to use the remote's default branch. Omitting `--branch` should have exactly this effect. Of course, no branch will actually be checked out yet, due to `--no-checkout`, but we'll be fetching the right things from the remote, and then we check out the appropriate branch later (either the specified branch, or the default remote branch as returned by `straight-vc-git--default-remote-branch`).

There might be something hiding in the interaction of overriding `:branch` inside of `:fork` or something like that, but I am fully comfortable leaving that for Later(tm).

**Comment by progfolio** [comment]
Thank you, @ahyatt!

**Comment by ahyatt** [comment]
Super happy to see this go in, thanks everyone for their patient support.

**Pull request merged**